pycozmo/window.py

```diff
@@ -5,159 +5,167 @@
 """
 
 import math
-from typing import Optional, Any
+from typing import Optional, List, Tuple, Any
+
+from . import exception
 
 
 __all__ = [
     "BaseWindow",
     "ReceiveWindow",
-    "SendWindowSlot",
     "SendWindow",
 ]
 
 
 class BaseWindow(object):
+    """ Base communication window class. """
 
-    def __init__(self, seq_bits: int, size: Optional[int] = None) -> None:
+    def __init__(self, seq_bits: int, size: Optional[int] = None, max_seq: Optional[int] = None) -> None:
+        """ Crate a window by specifying either sequence number bits or size of the window. """
         if size is None:
-            self.size = int(math.pow(2, seq_bits - 1))
-        elif size < 0 or size > int(math.pow(2, seq_bits - 1)):
+            if seq_bits < 1:
+                raise ValueError("Invalid sequence number bits.")
+            size = int(math.pow(2, seq_bits - 1))
+        elif size <= 0 or size > int(math.pow(2, seq_bits - 1)):
             raise ValueError("Invalid window size.")
-        else:
-            self.size = size
+        # Size of the window.
+        self.size = size
+        # Next expected sequence number (0, max_seq-1).
         self.expected_seq = 1
-        self.last_seq = 0
-        self.max_seq = int(math.pow(2, seq_bits)) - 1
+        # Maximum sequence number (first invalid).
+        self.max_seq = max_seq or int(math.pow(2, seq_bits))
+        if self.max_seq % self.size != 0:
+            raise ValueError("The maximum sequence number must be a multiple of the window size must.")
+        # Window data
+        self.window = [None for _ in range(self.size)]
```

Comment:
Have you thought about the advantages/disadvantages of using a dynamically allocated window vs the current pre-allocated one?

I wonder if it would be more clear and/or efficient to use the append/pop functions of lists instead of reassigning their values.

Reply:
Yes, this could work with the send window. The receive window will have to be a `dict`.

Originally it seemed more complicated to use `list`/`dict` for the windows. I was envisioning very different resend strategy though.

Reply:
That's right, for the receive window the current system is much more straight forward. I forgot that packets can come unsorted.

It makes more sense to keep it as it is now then, rather than have two completely different window types.